src/go/gcsrunner/fetch_config.go

```diff
@@ -27,52 +27,67 @@ import (
 	"cloud.google.com/go/storage"
 	"github.com/cenkalti/backoff"
 	"github.com/golang/glog"
+	"golang.org/x/oauth2/google"
 	"google.golang.org/api/option"
-
-	google_oauth "golang.org/x/oauth2/google"
 )
 
-// FetchConfigOptions provides a set of configurations when fetching and writing config files.
-type FetchConfigOptions struct {
-	BucketName                    string
-	ConfigFileName                string
-	WriteFilePath                 string
-	FetchGCSObjectInitialInterval time.Duration
-	FetchGCSObjectTimeout         time.Duration
-}
-
-var findDefaultCredentials = google_oauth.FindDefaultCredentials
+var (
+	// can be overwritten in unit tests
+	newGCS             = newGCSClient
+	newDefaultCredsGCS = newDefaultCredsGCSClient
+	osCreate           = func(p string) (io.WriteCloser, error) { return os.Create(p) }
+)
 
-type getObjectRequest struct {
-	Bucket, Object string
-}
-type storageObjectReader interface {
-	readObject(ctx context.Context, req getObjectRequest) (io.Reader, error)
+// can be mocked in unit tests
+type gcsReader interface {
+	Reader(ctx context.Context, bucket, object string) (io.Reader, error)
 }
 
-type gcsObjectReader struct {
+// Implements a minimal GCS client for reading files from GCS.
+type gcsClient struct {
 	client *storage.Client
 }
 
-func (g *gcsObjectReader) readObject(ctx context.Context, req getObjectRequest) (io.Reader, error) {
-	return g.client.Bucket(req.Bucket).Object(req.Object).NewReader(ctx)
+func (c *gcsClient) Reader(ctx context.Context, bucket, object string) (io.Reader, error) {
+	return c.client.Bucket(bucket).Object(object).NewReader(ctx)
 }
 
-var newStorageObjectReader = func(ctx context.Context, opts ...option.ClientOption) (storageObjectReader, error) {
-	c, err := storage.NewClient(ctx, opts...)
-	if err != nil {
-		return nil, err
-	}
-	return &gcsObjectReader{c}, nil
-}
-
-type file interface {
-	io.Writer
-	io.Closer
+// FetchConfigOptions provides a set of configurations when fetching and writing config files.
+type FetchConfigOptions struct {
+	// ServiceAccount, if provided, is impersonated in the GCS fetch call.
+	// If left blank, the default credentials are used.
+	ServiceAccount                string
+	BucketName                    string
+	ConfigFileName                string
+	WriteFilePath                 string
+	FetchGCSObjectInitialInterval time.Duration
+	FetchGCSObjectTimeout         time.Duration
 }
 
 var (
-	osCreate = func(p string) (file, error) { return os.Create(p) }
+	// can be overwritten in unit tests
+	newGCSClient = func(ctx context.Context, sa string) (gcsReader, error) {
```

Comment:
optional: Instead of making these function vars, they could just be functions e.g

```
var (
 gcsCient = newGCSClient
)
func newGCSClient(...)...
```